crates/sui-core/src/authority.rs

```diff
@@ -678,13 +678,46 @@ impl AuthorityState {
         certificate: &CertifiedTransaction,
         transaction_digest: TransactionDigest,
     ) -> SuiResult<(InnerTemporaryStore, SignedTransactionEffects)> {
+        let (inner_temp_store, effects, _execution_error) = self
+            .exec_transaction(certificate, transaction_digest)
+            .await?;
+        // TODO: Distribute gas charge and rebate, which can be retrieved from effects.
+        let signed_effects = effects.to_sign_effects(self.epoch(), &self.name, &*self.secret);
+
+        Ok((inner_temp_store, signed_effects))
+    }
+
+    pub async fn call_transaction(
+        &self,
+        transaction: &Transaction,
+        transaction_digest: TransactionDigest,
+    ) -> Result<SuiTransactionEffects, anyhow::Error> {
+        transaction.verify()?;
+        let (_inner_temp_store, effects, _execution_error) = self
+            .exec_transaction(transaction, transaction_digest)
+            .await?;
+        SuiTransactionEffects::try_from(effects, self.module_cache.as_ref())
+    }
+
+    async fn exec_transaction<T>(
+        &self,
+        transaction: &TransactionEnvelope<T>,
+        transaction_digest: TransactionDigest,
+    ) -> Result<
+        (
+            InnerTemporaryStore,
+            TransactionEffects,
+            Option<ExecutionError>,
+        ),
+        SuiError,
+    > {
         let (gas_status, input_objects) =
-            transaction_input_checker::check_transaction_input(&self.database, certificate).await?;
+            transaction_input_checker::check_transaction_input(&self.database, transaction).await?;
 
         // At this point we need to check if any shared objects need locks,
         // and whether they have them.
         let shared_object_refs = input_objects.filter_shared_objects();
-        if !shared_object_refs.is_empty() && !certificate.signed_data.data.kind.is_system_tx() {
+        if !shared_object_refs.is_empty() && !transaction.signed_data.data.kind.is_system_tx() {
```

Comment:
> If the transaction contains shared objects, we need to ensure they have been scheduled
            // for processing by the consensus protocol.

Does this also apply to `call_transaction`? If so, it seems to open up a vulnerability for DDOS where consensus protocol can be called without gas. cc @lxfind

Reply:
iirc, `check_shared_locks` only checks local DB of AuthorityState, and the fullnode that processes this dry_run request will not send it elsewhere either, so it will not add extra traffic or workload to other nodes. 
https://github.com/gegaowp/sui/blob/5d346fdcceb3e0f42b2b23feb7cffd367a98ed49/crates/sui-core/src/authority.rs#L530
will leave it to Xun to confirm.

Reply:
Correct, this only checks for shared object locks, and does not do consensus.
However, if a transaction does involve shared objects, trying to execute it here without consensus is not going to work.
What do we want to do with call_transaction that involves shared objects? Maybe we should abort it earlier in call_transaction by checking this first.

Reply:
@lxfind thanks for the info!
> trying to execute it here without consensus is not going to work.

other than the sequence number, does the execution also need anything else from consensus?

> What do we want to do with call_transaction that involves shared objects?

the original ideal it to implement an eth_call equivalent so ideally we can dry run txns with shared objects as well, but if that is indeed tricky, I think it would still be useful for owned objects only. And yes I should abort earlier if do so.

Reply:
Sequence number is what it needs, which is usually set by the consensus.
One alternative is that you could just use whatever is the latest state of the shared objects to execute the transaction, which may be useful in most cases if the shared object is not updated all the time. To do that, you can probably bypass the shared object lock check if it's attempting to execute a transaction instead of a certificate.
Some tests would be useful just to make sure this works.

Reply:
@lxfind I made a couple of changes accordingly and it should be ready for another review, thanks!
- dry run will skip the shared objects check, I also reverted the helper function change as code sharing is small now
- added a dry run test.